art/defences/trainer/adversarial_trainer_oaat_pytorch.py

```diff
@@ -0,0 +1,1224 @@
+        self, x: "torch.Tensor", perturbation: "torch.Tensor", eps_step: Union[int, float, np.ndarray]
+    ) -> "torch.Tensor":
+        """
+        Apply perturbation on examples.
+
+        :param x: Current adversarial examples.
+        :param perturbation: Current perturbations.
+        :param eps_step: Attack step size (input variation) at each iteration.
+        :return: Adversarial examples.
+        """
+        import torch
+
+        eps_step = np.array(eps_step, dtype=ART_NUMPY_DTYPE)
+        perturbation_step = torch.tensor(eps_step).to(self._classifier.device) * perturbation
+        perturbation_step[torch.isnan(perturbation_step)] = 0
+        x = x + perturbation_step
+        if self._classifier.clip_values is not None:
+            clip_min, clip_max = self._classifier.clip_values
+            x = torch.max(
+                torch.min(x, torch.tensor(clip_max).to(self._classifier.device)),
+                torch.tensor(clip_min).to(self._classifier.device),
+            )
+
+        return x
+
+    def _projection(
+        self, values: "torch.Tensor", eps: Union[int, float, np.ndarray], norm_p: Union[int, float, str]
+    ) -> "torch.Tensor":
+        """
+        Project `values` on the L_p norm ball of size `eps`.
+
+        :param values: Values to clip.
+        :param eps: Maximum norm allowed.
+        :param norm_p: L_p norm to use for clipping supporting 1, 2, `np.Inf` and "inf".
+        :return: Values of `values` after projection.
+        """
+        import torch
+
+        # Pick a small scalar to avoid division by 0
+        tol = 10e-8
+        values_tmp = values.reshape(values.shape[0], -1)
+
+        if norm_p == 2:
+            if isinstance(eps, np.ndarray):
+                raise NotImplementedError(
+                    "The parameter `eps` of type `np.ndarray` is not supported to use with norm 2."
+                )
+
+            values_tmp = (
+                values_tmp
+                * torch.min(
+                    torch.tensor([1.0], dtype=torch.float32).to(self._classifier.device),
+                    eps / (torch.norm(values_tmp, p=2, dim=1) + tol),
+                ).unsqueeze_(-1)
+            )
+
+        elif norm_p == 1:
+            if isinstance(eps, np.ndarray):
+                raise NotImplementedError(
+                    "The parameter `eps` of type `np.ndarray` is not supported to use with norm 1."
+                )
+
+            values_tmp = (
+                values_tmp
+                * torch.min(
+                    torch.tensor([1.0], dtype=torch.float32).to(self._classifier.device),
+                    eps / (torch.norm(values_tmp, p=1, dim=1) + tol),
+                ).unsqueeze_(-1)
+            )
+
+        elif norm_p in [np.inf, "inf"]:
+            if isinstance(eps, np.ndarray):
+                eps = eps * np.ones_like(values.cpu())
+                eps = eps.reshape([eps.shape[0], -1])  # type: ignore
```

Comment:
Is this type ignore necessary?

Reply:
Yes, otherwise mypy gives this error:

art/defences/trainer/adversarial_trainer_oaat_pytorch.py:1205: error: Item "int" of "Union[int, float, ndarray[Any, Any]]" has no attribute "reshape"
art/defences/trainer/adversarial_trainer_oaat_pytorch.py:1205: error: Item "float" of "Union[int, float, ndarray[Any, Any]]" has no attribute "reshape"
art/defences/trainer/adversarial_trainer_oaat_pytorch.py:1205: error: Item "int" of "Union[int, float, ndarray[Any, Any]]" has no attribute "shape"
art/defences/trainer/adversarial_trainer_oaat_pytorch.py:1205: error: Item "float" of "Union[int, float, ndarray[Any, Any]]" has no attribute "shape"

Reply:
Corrected it and removed the type ignore